### DISCUSSION TITLE
Bring Your Own Repo Discussion

### DISCUSSION THREAD

**Discussion by eliotwrobson** [PrairieLearn/PrairieLearn#6144] [Dev]
A thread for discussing technical details of the BYO repo plan @nwalters512 .

**Comment by nwalters512** [comment]
Here's roughly what I have in mind:

- There would be a PrairieLearn GitHub app. Users could authenticate with GitHub and then install the app on either a personal account or an organization as a whole.
- Once the app is installed, users could select a repo for their course from a course admin page.
  - We'd need to figure out what the process of cutting over from one repo to another is. Do we rely on folks syncing all files between them out-of-band? Should we offer to force-push any existing contents to the new repo?
- We'd need to update our syncing code that does git operations (`clone`/`pull`/`push`) to use credentials that we'd presumably get from GitHub (I've never built a GitHub integration before, so I'm not 100% sure how that works).
- Wherever relevant, we should design this with other source control platforms in mind. While we should only support GitHub in the first iteration of this, we shouldn't make any choices that would make it difficult for us to expand to build integrations for GitLab/Bitbucket/etc. in the future.

@mwest1066 when we had discussed BYOR in the past, we threw around the idea of this being an enterprise-only feature (the assumption being that this feature would appeal most to either really large institutions, or to us as the operators of large multi-tenant installations). What are your current thoughts about that?

**Reply by eliotwrobson** [reply]
Small disclaimer that I'm not an experienced web developer, nor a super expert on how PL has it's permissions set up / does git operations.

So after doing some reading, it seems like having a Github app might not be the best choice for this? Based on [this](https://docs.github.com/en/developers/apps/getting-started-with-apps/about-apps#determining-which-integration-to-build) page, it seems like you wouldn't want any theoretical PL Github app to act of its own volition, rather you want the ability for individual users to be able to edit / view course files. In that case, you could go with a simpler design that requires users to provide PL with some type of access token, so that PL can perform actions on the Git repo on their behalf. I think this would also have the added benefit of less coupling with Github.

I think it's reasonable to allow force-pushing the contents to a new repo as a feature.

**Reply by nwalters512** [reply]
While that would work, I think it has a few downsides:

- Either everyone who wishes to do *anything* on PL (syncing, editing in the browser, etc.) would have to provide a token, or we would have to delegate all actions to a single user's token. The former introduces a fair amount of friction for users; the latter introduces a single point of failure for a course. The former also has the downside that it makes it difficult for support staff (think institution admins or PrairieLearn admins) to interact with course content: they'd either be unable to take any actions, or they'd have to use another user's credentials, which feels weird.
- There's no way to limit token access. So, a given token would allow PrairieLearn access to *any* repo that the user has access to. Frankly, we do not want that.

I actually think that the mental model of "PrairieLearn as a whole interacts with your course repo" is a pretty good one. It's how PL works today (and this has worked fine), and opens us up to some neat possibilities: think PL automatically opening PRs with a migration to Bootstrap 5, or maybe maintaining automatically setting up webhooks to sync course content when new commits are pushed to the main branch.

I may be wrong though, let's talk this though some more!

Aside: another possibility is a "sign in with GitHub" button. Then we don't have to worry about folks copy/pasting access tokens, or those tokens expiring. That still has the aforementioned downsides though (I think).

**Reply by eliotwrobson** [reply]
So I think you're right that "sign in with GitHub" is probably easier than having users manually generate and provide a token (I think this is actually the same as the OAuth solution mentioned on that page). I think that I was thinking about this from the standpoint of having to build something from the ground up, rather than the easiest thing to build out from existing infrastructure. 

With the first issue, everyone that wanted to interact with PL in a BYOR would need a login to Github and access to the particular course repo. While using a separate app avoids this, this also brings in significant overhead in terms of needing to maintain two different applications. Also, I actually haven't seen a Github app on the marketplace that can listen for requests from a separate application (it seems like everything needs to run internally somehow, but I'm by no means an expert). I also don't think the OAuth solution introduces massive overhead, given that this is a situation where the repo is being "self-managed".

I guess what I'm saying is that I think both solutions have their downsides, and actually the use of a GitHub app is the most natural given the existing infrastructure, but I'm not sure that the feasibility issues outweigh the downsides you mention.

**Reply by eliotwrobson** [reply]
After reading about this some more [here](https://docs.github.com/en/developers/apps/guides/using-the-github-api-in-your-app), it seems like individual GitHub apps can listen to webhooks through the GitHub API. But I think going through that is going to require using OAuth or similar authentication tokens anyway.

**Reply by echuber2** [reply]
We made a simple GitHub App for the Illinois "repo creator" service that targets our orgs on the cloud enterprise service. I think the GitHub App model does work for your use case:

> We'd need to update our syncing code that does git operations (clone/pull/push) to use credentials that we'd presumably get from GitHub (I've never built a GitHub integration before, so I'm not 100% sure how that works).

A GitHub App, from GitHub's perspective, is really just a named token with a set of permissions attached. When people install your App on their org, they give it access to do the things it's permissioned for. So in this case, a PL GitHub App would request the permission to do operations of various kinds on the repos in an org, and when somebody elects to add your App, they grant it those permissions.

**Reply by echuber2** [reply]
However it may also work to just have people generate a "deploy key" for their repo with write permissions, and have them enter that key in the PL admin screen somewhere. This is a key permissioned to a single repo (not associated with a specific user account). This would also have the benefit of not needing a special "Select your org" tool like a GitHub App would; whereas the App would have to query what it's been granted access to and expose only the right setting to the PL admin user through some selector (which is risky in that it also requires verifying the admin user's GitHub associated identity somehow), instead you could just tell the PL admin user "drop your deploy key here." https://docs.github.com/en/developers/overview/managing-deploy-keys#deploy-keys

**Reply by echuber2** [reply]
^Also, GitLab and BitBucket both have deploy key features similar to GitHub's, so this would be extensible. (BitBucket calls them "access keys.")

https://docs.gitlab.com/ee/user/project/deploy_keys/
https://confluence.atlassian.com/bitbucketserver/ssh-access-keys-for-system-use-776639781.html

**Reply by ffund** [reply]
My highly non-technical two cents: I like the deploy key approach. Apps make me nervous, I feel like I don't fully understand what permissions I am giving them, or how to remove permissions. Whereas a deploy key (+ webhook to sync content when I push to the repo) "feels" straightforward and transparent.

**Reply by ffund** [reply]
I generally prefer something that is at the repo level and not account/organization level. (I understand that you can specify which repos an app has permission for, but I still don't like that it's installed on account/organization.)

**Reply by nwalters512** [reply]
Ah!! I completely forgot about deploy keys (and even then I forgot that you can configure them with write permissions). That might be ideal. We could even have PL generate the keypair and give the user the public key to add to their source control service. Excellent idea @echuber2.

**Reply by eliotwrobson** [reply]
Looking around, deploy keys seem like a really flexible solution and very feasible to implement given the current infrastructure of PL. Not to get too far ahead of ourselves, but if that's the route we want to go, what kind of next steps should we take? Draw up design docs of some type?

**Reply by nwalters512** [reply]
That seems reasonable, yeah! We'd typically use an [RFC](https://github.com/PrairieLearn/PrairieLearn/tree/master/rfcs) for discussing new things like this. Don't feel like you have to copy the structure of the existing ones if you don't want to; just getting stuff written down into a form that we can discuss and iterate on (PR comments are great!) is the important bit.

**Reply by eliotwrobson** [reply]
Sounds good! We'll get started as soon as we get everyone onboarded on our dev team.

**Comment by nwalters512** [comment]
This discussion feels like it's related to the discussion of using GitHub submodules that @jesus-villalobos brought up on Slack. That would be tricky with PL's current model of "machine user that can access every single repo" since course A could declare course B as a submodule and get access to its contents. But if we were working with user-based credentials, we'd get access to everything the user has access to, making submodules much easier to implement. Notably, this would not be super straightforward with deploy tokens, unless you collected multiple deploy tokens (one for the main repo, one per submodule).

**Reply by echuber2** [reply]
Related discussion: https://github.com/PrairieLearn/PrairieLearn/discussions/5907

There are potential issues with submodules but also several alternatives to them, so I'll just leave some resources here:

- Why submodules are bad (probably one of many similar articles)
  https://codingkilledthecat.wordpress.com/2012/04/28/why-your-company-shouldnt-use-git-submodules/

# Alternative: git subtree

- git subtree tool (originally a community contribution)
  https://manpages.debian.org/testing/git-man/git-subtree.1.en.html
- subtree merge strategy for git (official and fully manual)
  https://git-scm.com/book/en/v2/Git-Tools-Advanced-Merging#_subtree_merge

More info about subtree (tool or strategy)

- https://stackoverflow.com/questions/32407634/when-to-use-git-subtree
- https://www.atlassian.com/git/tutorials/git-subtree
- https://medium.com/@porteneuve/mastering-git-subtrees-943d29a798ec

# Other alternatives

I'm not sure of the relative strengths of these tools. Some of these predate the subtree option.

- Google Repo tool (from Android tooling)
  https://source.android.com/docs/setup/develop/repo
- braid
  https://github.com/cristibalan/braid
- git-subrepo
  https://github.com/ingydotnet/git-subrepo
